torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -599,7 +599,7 @@ def _root_pre_forward(
 
         input_dtype: Optional[torch.dtype] = state.mixed_precision.param_dtype
 
-        should_cast_forward_inputs = all(
+        should_cast_forward_inputs = len(state._handles) > 0 and all(
```

Comment:
reverted this change in #100290 as it breaks some use cases, it is very common that root module does not have parameters, but it still wants to cast forward inputs in default without customizing children's mixed precison config

Reply:
I think the original motivating issue https://github.com/pytorch/pytorch/issues/99545 is legit. We should think about what the right behavior should be, and if there are cases that cannot be supported, what we can do to remediate.

For the original issue, we may be able to fix that case separately since it relates to `model.eval()` mode and forcing full precision. However, we should be careful to not fix that as one-off and make things more confusing. I will think about this more next week.

Reply:
Yeah, this PR explicitly required users to set `cast_forward_inputs=False` if they wanted to avoid hitting #99545 while using an FSDP root module with no direct parameters so if that is a sufficiently common scenario, a broader solution @awgu alludes to definitely makes sense! (thanks @zhaojuanmao!)

To facilitate testing that solution (and per @rohan-varma's request) I'm adding a test set in https://github.com/pytorch/pytorch/pull/100349 that should validate a range of cast forward scenarios (described in the PR) with whatever
broader solution to #99545 the team decides on implementing (with one of many potential candidate solutions listed as a file suggestion).

Reply:
just had a second though, I think "len(state.handles) > 0" check works for auto wrapping case, but not work well for manual wrapping case well, as 'state.handles' in root modules does not include handles in the children instances, it will confuse current FSDP mixed precision design. 

If it is mainly to handle model.eval() case, would suggest to simply check whether the model is in .eval() mode, if so, globally disable mixed precision. @awgu @rohan-varma

Reply:
Makes sense to me! While there are many potential solutions, that's effectively the one I included [as a suggestion](https://github.com/pytorch/pytorch/pull/100349#discussion_r1181312134) on my [test set PR](https://github.com/pytorch/pytorch/pull/100349) yesterday (i.e. only considering casting in root if we're in training mode, mp would be disabled in child modules by each module's `_force_full_precision` property)

Reply:
Any chance the team has decided on the preferred approach to fully accommodate `model.eval()`-mode forced full precision (currently failing for root modules with no direct handles as discussed above)? I'm sure there are many higher priority issues at the moment, just wanted to make sure it didn't fall off the radar as I know could easily happen! Thanks again for all your work!
@awgu @zhaojuanmao

Reply:
Hi @speediedan. Sorry for the long delays on this. We still need to discuss this design. Is there any way to work around this for you locally (even if it is hacky), or are you blocked on this?

Reply:
No worries @awgu, thanks for responding, I know the team has a lot on its plate!

Do you anticipate the refactored design making it into a nightly anytime soon/the `2.1.0` release? It's not urgent from my perspective. If it is expected to be more than another week or two, I should be able to add a temporary workaround for [the downstream package](https://finetuning-scheduler.readthedocs.io/en/latest) I maintain. Thanks!